crates/bevy_render/src/view/visibility/mod.rs

```diff
@@ -139,72 +142,70 @@ pub fn update_frusta<T: Component + CameraProjection + Send + Sync + 'static>(
-    mut visible_entity_query: ParamSet<(
-        Query<&mut ComputedVisibility>,
-        Query<(
-            Entity,
-            &Visibility,
-            &mut ComputedVisibility,
-            Option<&RenderLayers>,
-            Option<&Aabb>,
-            Option<&NoFrustumCulling>,
-            Option<&GlobalTransform>,
-        )>,
+    mut visible_entity_query: Query<(
+        Entity,
+        &Visibility,
+        &mut ComputedVisibility,
+        Option<&RenderLayers>,
+        Option<&Aabb>,
+        Option<&NoFrustumCulling>,
+        Option<&GlobalTransform>,
     )>,
 ) {
-    // Reset the computed visibility to false
-    for mut computed_visibility in visible_entity_query.p0().iter_mut() {
-        computed_visibility.is_visible = false;
-    }
-
     for (mut visible_entities, frustum, maybe_view_mask) in view_query.iter_mut() {
         visible_entities.entities.clear();
         let view_mask = maybe_view_mask.copied().unwrap_or_default();
 
-        for (
-            entity,
-            visibility,
-            mut computed_visibility,
-            maybe_entity_mask,
-            maybe_aabb,
-            maybe_no_frustum_culling,
-            maybe_transform,
-        ) in visible_entity_query.p1().iter_mut()
-        {
-            if !visibility.is_visible {
-                continue;
-            }
-
-            let entity_mask = maybe_entity_mask.copied().unwrap_or_default();
-            if !view_mask.intersects(&entity_mask) {
-                continue;
-            }
-
-            // If we have an aabb and transform, do frustum culling
-            if let (Some(model_aabb), None, Some(transform)) =
-                (maybe_aabb, maybe_no_frustum_culling, maybe_transform)
-            {
-                let model = transform.compute_matrix();
-                let model_sphere = Sphere {
-                    center: model.transform_point3a(model_aabb.center),
-                    radius: (Vec3A::from(transform.scale) * model_aabb.half_extents).length(),
-                };
-                // Do quick sphere-based frustum culling
-                if !frustum.intersects_sphere(&model_sphere, false) {
-                    continue;
+        let viz = Arc::new(RwLock::new(Vec::new()));
+
+        visible_entity_query.par_for_each_mut(
+            &pool,
+            32,
+            move |(
+                entity,
+                visibility,
+                mut computed_visibility,
+                maybe_entity_mask,
+                maybe_aabb,
+                maybe_no_frustum_culling,
+                maybe_transform,
+            )| {
+                computed_visibility.is_visible = false;
+                if visibility.is_visible {
+                    let entity_mask = maybe_entity_mask.copied().unwrap_or_default();
+                    if view_mask.intersects(&entity_mask) {
+                        let mut skip = false;
```

Comment:
Wouldn't replacing the `continue`s with `return`s be better here, instead of introducing an bool?

This change means that below, `intersects_obb` is always called, even the `intersects_sphere` call already set skip to true.

Reply:
Good catch. As far as I'm aware, I can't use return or continue from within a closure, right?

Reply:
You can't return from a closure in the sense that the return won't exit the scope above the closure, but if the return is to end the closure early you should be able to.

```rs
fn main() {
    let foo = |i| {
        if i % 2 == 0 {
			// this will stop the execution of foo, but main will not exit
            return;
        } else {
            println!("{i}");
        }
        println!("end");
    };
    for i in 0..10 {
        foo(i);
    }
}
```

Reply:
Ah, duh, that makes perfect sense. TIL.